Add chat rich embeds

MatinAniss opened the pull request:
Adds chat rich embeds for twitch clips.

Example screenshots:
Night theme:
<img width="964" alt="Screen Shot 2023-04-08 at 5 08 05 am" src="https://user-images.githubusercontent.com/76515905/230664031-af5a0aac-46ff-4d13-8750-1c34beb492fb.png">

Light theme:
<img width="964" alt="Screen Shot 2023-04-08 at 5 08 33 am" src="https://user-images.githubusercontent.com/76515905/230664040-4751dae6-037a-4704-aa27-b718216ef8ef.png">

Closes #360

AnatoleAM reviewed (commented):
After some testing, it seems that the rich embed does not apply if the message doesn't come from the tab you just typed in.

AnatoleAM commented on src/site/twitch.tv/modules/chat/ChatPubSub.vue:
```diff
@@ -103,6 +107,25 @@ async function onModerationAction(msg: PubSubMessageData.ModAction) {
+	if (!ctx) return;
+
+	// Find the message
+	const message = messages.messageById(msg.message_id);
```
This is most likely why, use `messages.awaitMessage` instead.

MatinAniss replied:
I am using `messages.messageById` because the chat message arrives before the pubsub message does (on the sender), but this is the opposite (on the receiver).

AnatoleAM commented on src/composable/chat/useChatMessages.ts:
```diff
@@ -250,6 +250,21 @@ export function useChatMessages(ctx: ChannelContext) {
+	 * @param id the message id
+	 * @returns message
+	 */
+	function messageById(id: string): ChatMessage | void {
```
This method can be removed, as in almost any case where it might be useful, the existing `awaitMessage` would cover it.

AnatoleAM commented on src/site/twitch.tv/modules/chat/components/message/UserMessage.vue:
```diff
@@ -71,6 +71,9 @@
 		</span>
 
+		<!-- Chat Rich Embed -->
+		<RichEmbed v-if="msg.richEmbed.request_url" :richEmbed="msg.richEmbed" />
```
Lint issue here

```suggestion
		<RichEmbed v-if="msg.richEmbed.request_url" :rich-embed="msg.richEmbed" />
```

AnatoleAM reviewed (changes requested):
Overall this is nice, just these fixes, and then add a changelog entry.

MatinAniss commented:
- Fixed issue where rich embed only appeared on the senders side.
- Added check in awaitMessage function to check if the message has already been pushed to the displayed or buffer (this was done because the pubsub message arrives after the message arrives which results in the promise waiting for something that has already happend).
- Fixed vue/attribute-hyphenation lint issue.
- Added the changelog entry to nightly.

AnatoleAM commented on CHANGELOG-nightly.md:
```diff
@@ -1,6 +1,10 @@
+### Version 3.0.3.2000
```
Rebase or merge master, then have this under Untitled Version

AnatoleAM commented on src/composable/chat/useChatMessages.ts:
```diff
@@ -258,6 +258,17 @@ export function useChatMessages(ctx: ChannelContext) {
+			const displayedMessage = data.displayed.find((msg) => {
+				return msg.id === id;
+			});
+			if (displayedMessage) resolve(displayedMessage);
```
You should return here, otherwise the code below will still run

```suggestion
			if (displayedMessage) return resolve(displayedMessage);
```

AnatoleAM reviewed (commented):
Looking a lot more functional!

MatinAniss commented on src/site/twitch.tv/modules/chat/components/message/RichEmbed.vue:
```diff
@@ -0,0 +1,69 @@
+
+defineProps<{
+	richEmbed: RichEmbed;
+}>();
```
I don't know the possible twitch embeds as it is all undocumented, I added a condition to check if it is a clip.

MatinAniss commented:
- Updated CHANGELOG-nightly.md.
- Added the returns for the resolves to stop the function from continuing unnecessarily.
- Fixed the log debug message.
- Fixed the styling AnatoleAM suggested.
- Added a condition to check if the embed is a clip which adds the "Clipped by ".

Melonify commented on src/site/twitch.tv/modules/chat/ChatPubSub.vue:
```diff
@@ -103,6 +107,27 @@ async function onModerationAction(msg: PubSubMessageData.ModAction) {
+	});
+	if (!message) return;
+
+	const rmsg = ref(message);
```
Don't think setting a `ref` here is necessary, this won't really do anything, creating refs like this doesn't introduce reactivity, and the message is already reactive.

Other than that, this PR looks good.

AnatoleAM reviewed (approved):
Thank you for the contribution! Drop your twitch username if you'd like the contributor badge 🫡

MatinAniss commented:
Twitch username: MatinAniss
:)

AnatoleAM commented:
@MatinAniss entitled :slightly_smiling_face: 

also, via #472 this feature is live right now on the nightly version!
